src/core/server/opensearch_data/opensearch_data_service.ts

```diff
@@ -0,0 +1,98 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+import { Observable, Subject } from 'rxjs';
+import { first, map, shareReplay, takeUntil } from 'rxjs/operators';
+import { CoreService } from 'src/core/types';
+import { AuditorFactory, AuditTrailStart } from '../audit_trail';
+import { CoreContext } from '../core_context';
+import { Logger } from '../logging';
+import { OpenSearchClientConfig } from '../opensearch/client';
+import { OpenSearchConfig, OpenSearchConfigType } from '../opensearch/opensearch_config';
+import { OpenSearchService } from '../opensearch/opensearch_service';
+import {
+  InternalOpenSearchServiceSetup,
+  InternalOpenSearchServiceStart,
+} from '../opensearch/types';
+import { pollOpenSearchNodesVersion } from '../opensearch/version_check/ensure_opensearch_version';
+import {
+  InternalSavedObjectsServiceSetup,
+  InternalSavedObjectsServiceStart,
+  SavedObjectsClient,
+} from '../saved_objects';
+import { SavedObjectsClientContract } from '../types';
+import { DataSourceClient } from './client/data_source_client';
+
+interface StartDeps {
+  savedObjects: InternalSavedObjectsServiceStart;
+  auditTrail: AuditTrailStart;
+}
+
+export class OpenSearchDataService implements CoreService<any, any> {
+  private readonly log: Logger;
+  private readonly config$: Observable<OpenSearchConfig>;
+  private auditorFactory?: AuditorFactory;
+  private stop$ = new Subject();
+  private opensearchDashboardsVersion: string;
+  private savedObjectClient?: SavedObjectsClientContract;
+  private dataSourceClient?: DataSourceClient;
+
+  constructor(private readonly coreContext: CoreContext) {
+    this.opensearchDashboardsVersion = coreContext.env.packageInfo.version;
+    this.log = coreContext.logger.get('opensearch-data-service');
+    this.config$ = coreContext.configService
+      .atPath<OpenSearchConfigType>('opensearch') // TODO: update if we'll add data source specific configs
+      .pipe(map((rawConfig) => new OpenSearchConfig(rawConfig)));
+  }
+
+  public async setup(): Promise<any> {
+    this.log.debug('Setting up opensearch data service');
+
+    const config = await this.config$.pipe(first()).toPromise();
+
+    // TODO: update accordingly when we decide how to check node/version compatibility in setup stage
```

Comment:
with multiple datasource, check opensearch version compatibility when service start is not sufficient. 
we will need to check compatibility for every connection to destination opensearch endpoint

Reply:
Yes, it means to check compatibility of every existing dataSource, and that's why we can't reuse the old code. I'll make a note of it, or create an issue to track.